### ISSUE TITLE
[Feature]: Bat mode overhaul

### ISSUE THREAD

**Issue by MarioSMB** [TeamLapen/Vampirism#1160]
### What?

I appreciate that the following proposed design would be ambitious and challenging to achieve, but please bear with me as this would revolutionize one of the more notable aspects of vampirism:

### Activation
Instead of needing to cycle through the ability wheel, the player would simply need to double jump to activate bat mode - just as one might do to activate flight in creative mode.  
The player would be restored to regular form upon touching the ground. Double jumping would not toggle it off midair.

### Flight
The thought of needing a custom flight system may seem a bit daunting, but there is one way to break it down a little - the goal is essentially what some mods would implement as a jetpack; holding jump goes up, decent vertical speed with very little horizontal speed, releasing jump falls back down.  
With the refined controls this would allow, flying as a vampire would feel intuitive and enjoyable while allowing a more refined balance, focus being moved from long distance travel to quick escapes and reaching high places.

### Exhaustion
There is a timer that exists currently to limit the amount of time one can spend in bat form, which would still be used - possibly with a default allowing 3-5 seconds of flight.  
However, instead of transforming the player back when reaching the time limit, this would instead be used to mark the bat's "exhausted" state, where jump no longer allows vertical travel, instead helping the bat glide down to the ground - which as mentioned, upon touching ground would transform the player back into regular form.  
Flight would also be taxing on thirst (possibly to the same degree as sprinting), the exact balance of which could be decided - it may be fine as it is now.

### Conclusion
Thank you for reading, and I hope some consideration can be made to take an approach like this! I love Vampirism and truly wish to see it improve, creative mode flight has always been a weak point to me for its impact on gameplay.

### Mod experience

Bat mode is probably the most underwhelming "cheese" ability in the entire vampire arsenal; flight in the form of creative mode makes for a poor experience gameplay and balance wise.  
With the proposed design, bat mode would be an incredibly useful tool for escaping tight spots and reaching high places, no longer being just a long distance travel button - it would actually feel like a part of the moveset of a vampire.

### Contributions

Given the scope of these changes, I wish I could be of aid, but unfortunately I can only provide suggestions at this stage.  
If any testing or balance feedback is required, please let me know!

**Comment by maxanier** [comment]
## Activation
Maybe. I think having to manually select with the action wheel or a keybinding is reasonable. A player may double jump accidentally or on purpose (maybe for some other mod)

## Flight
Sounds good.
Additionally, we could limit maximum flight height to the surface level ´+ <amount>´ blocks

## Exhaustion
Yes, but I would have a higher default duration. Rather 10 to 15 seconds. 
Also, maybe we won't completely prevent ascend or horizontal acceleration, but just limit it to a much slower value. Or phrasing it the other way around. The player will have a short initial speed boost as a vampire

**Comment by maxanier** [comment]
@supermike1999
Yes, it might not be "realistic". But it does not have to be. It's mainly an idea we will have to test to see if it improves gameplay/feeling. Might be a stupid idea

**Comment by supermike1999** [comment]
@maxanier I didn’t mean to try and sound realistic. I didn’t explain very well on what I meant, but I get what you mean for gameplay improvements 👍

**Comment by lunofe** [comment]
Well, I think some of the approaches of this suggestion are interesting, but it's hard to think of this as an "overhaul" rather than a  nerf and quality-of-life deterioration. Let me go over some points.

---

### Double jump activation
Sounds interesting to me, but could be annoying for some players and will most definitely interfere with other mods actually implementing a double jump / dash ability and well, even creative mode for that sake, so while it surely could be a future option, it shouldn't replace the current way to activate bat mode by default I think.
<br>

### Non-creative-like flight system
I can't help but to disagree with this particular part:

> With the refined controls this would allow, flying as a vampire would feel intuitive and enjoyable [...]

It's of course personal preference, but while holding space to go up and releasing it to go down surely is intuitive, I don't think it would refine the controls at all. As a matter of fact, I'd say it achieves the exact opposite. The creative-like flight system allows for perfect directional control and even minimal adjustments of the height, while repeatedly holding & releasing the space bar to _somewhat_ stay on the same height is annoying at best, while also reducing the precise control the creative-like system had. The "_good old_" jetpack controls, are rather old than good and in my opinion, would feel like a leftover from the jetpack-era of 1.7.10 that no one wanted to rework yet.

I've tried using a jetpack with the maximum speed roughly set to the same as bat mode, and yes it works, but my hand now hurts. Seriously, try it, it's not pleasant.

https://user-images.githubusercontent.com/42828197/209664117-4bec3895-655d-4db8-b171-5964e6692809.mp4

<br>

### Limiting the height
It of course would make sense that bats can't fly hundreds if not thousands of blocks high, but I see two and a half problems with this:
- It's Minecraft. There may be a build limit, but no travel limit for height. Players are used to traveling way higher than the build limit for long distances with elytras and rockets or the riptide enchantment of tridents.
- In 1.18, the world generation has changed drastically. Mountains are way higher now, caves and ravines way deeper. Why can't I fly higher than 30 blocks in the plains biome, while I can fly 150 blocks higher (from that perspective) on top of a mountain? It's hard to find a relative value that would make sense.
   - What would happen in an edge case, say, I fly past a cliff? Would I just glide down without the option to fly up again? What happens when I fly over a random ravine, now appearing every 100 blocks?
<br>

### Exhaustion
I think it's a fair point to make the bat mode at least a bit more "realistic" with some sort of exhaustion, however as Max already stated, 3-5 seconds is way too short, and even 10-15 isn't enough in my opinion, at least for a maxed-out vampire. Maybe 10-15 seconds could be the starting point at level 4, going up to at least a minute for level 14 - and alongside that, maybe a new skill increasing the duration even further.
As Max also said, I think it would be a bad idea to completely prevent bat mode once the exhaustion kicks in. A slower speed for horizontal travel seems fine, but only if there would be an initial speed boost, that then only slows down to the speed we already have now. Otherwise it wouldn't make sense to use bat mode at that point (see below).
<br>

### Balancing
> [...] a more refined balance, focus being moved from long distance travel to quick escapes [...]

Bat mode already isn't the best way to travel long distances and most players use an elytra & fireworks once they've reached the late-game. 

https://user-images.githubusercontent.com/42828197/209656251-1f2bfce9-cf9b-4038-a2e2-09e2bdbcbee6.mp4

Actually, bat mode is way closer to jump running than an elytra (incl. rockets). Taking the average speeds (7 for jump running, 10 for bat mode and 30 for elytra), the elytra is more than six times faster than the speed gain between jump running and using bat mode.

If you were to travel 1000 blocks, using the average speeds from before (and some other methods for better comparison), these would be the outcomes:
| Method | ~ Duration |
| - | - |
| Jump running | 2min 22sec |
| Boat (water) | 2min 5sec |
| Bat mode | 1min 40sec |
| Minecart (powered rails) | 1min 30sec |
| Horse (max speed) | 1min 8sec |
| Elytra | 45sec |
| Elytra (incl. rockets) | 33sec |
| Boat (ice) | 25sec |
| Boat (blue ice) | 14sec |
| Boat (blue ice) in nether | 2sec |

Considering that, I don't think using bat mode to travel long distances is a bad thing or needs exhaustion at all. When I wouldn't know Vampirism yet and hear of a "bat mode", I'd think it would be to travel, instead of just being some sort of skill for quick escapes. https://github.com/TeamLapen/Vampirism/issues/1165 would be a better option for that.
<br>

### Conclusion
In my opinion, the non-creative-like flight system is a quality-of-life deterioration, which feels oddly out of place for its 1.7.10 vibes and takes away the precise controls we have now.

There's not really a reason to balance bat mode by adding exhaustion and trying to shift its use from travelling distances, because it isn't overly strong. Hunters, and even humans for that matter have access to an elytra, a direct vanilla counterpart which is even stronger, and don't forget that bat mode also only works at night, and reduces your HP. The only advantage bat mode currently has is its ability to stay stationary in the air and having no cooldown, at the cost of merely being faster than jump running.

Even what might feel like a small toning down for improved realism could lead to players not using bat mode at all, instead switching to the elytra even for the tiniest distances, taking away one of the few real advantages vampires have.

**Comment by MarioSMB** [comment]
Thanks for the feedback.  
Indeed, there are some points that are always going to be subject to debate, such as the exhaustion times added as an example.

Elytra is generally going to be the best for air-based travel, but there's a good reason for that: it's *the* end-game item, the thing you get for "beating" Minecraft. It has a lot of downsides at that; you must remove your strongest armor piece and you're always at risk of slamming into a surface and dying (not to mention the Mario 64-esque controls), or of the Elytra itself breaking.  
With that in mind, giving players with vampirism creative-mode flight early on - which, at least during the nighttime, allows virtually infinite distance travelled at no resource cost, is rather offputting to the game's balance. Balance is subjective of course, but there's already been cases where desired changes couldn't be made simply because the mod is falling back on creative-mode flight which can't be modified (examples: being able to sprint in the air and double-jumping out of it to walk around as a bat which tends to look ridiculous).

Regarding double jumps; presumably other mods would cancel the jump action when overriding it so as to avoid overlapping with anything that adds double jump mechanics (i.e. creative mode flight), but the concern of accidental double-jump presses is fair.

Limiting flying height isn't a bad idea, and would work great as an alternative to a shorter flying time limiting the overall distance one can travel. Maybe not needed, but the option is always nice!

For what it's worth, the video clip showing jetpack-like flight actually looks really good (besides the flames of course!). Spamming the jump button to stay at the same height might be a bit excessive indeed, perhaps crouch could be used to "hover"?  
Realism isn't the goal, just aiming for a more enjoyable experience as a vampire.  


In short, I don't believe the Elytra is a fair comparison to bat mode, as they *should* serve very different purposes and not make either obsolete.

**Comment by Piklach** [comment]
Just to follow on from what has been said. 
I think zins points regarding the travel speed are incredibly valid as most players reach vampire level 14 / Lord levels after about 4-5 days on average if they are intent on pushing through the levels. People can complete the game and kill the dragon within an hour or so if they are looking for quicker modes of transport such as the elyra. I think this invalidates the issue of balance between bat mode and other forms of transport. I feel as though adding a limit to flight would be abit annoying for some users as I know alot use flight for creating builds. 

I do disagree about a height limit too, I see no purpose in using a height limit for flight mode as it seems pointless to add for little to no reason. As we have discussed, elytras are much more efficient and movement and elytras also have no max height. I see no reason why bat mode should include a max height when there is no negative benefit to leaving it. 

On the topic of the jetpack like flight mode, I agree with zin that it seems irrelevant to add and seems to almost achieve the opposite effect of what you were desiring. Not only does it currently follow the phrase of "If it isn't broke, don't fix it". Adding this would effectively turn the effectiveness of flight mode backwards.

While I personally see like the idea of the activation you suggested, I feel it would be better to be an optional secondary method of activation (e.g. something toggleable in settings on a user to user basis) as to limit any crossover between multiple mods. Even if it is added, I would still opt to make the current method the default method. 

In conclusion. (No offence to yourself).  I see the majority of this idea as not only useless in terms of both time spent to implement but also the user experience in game. Not only is it not beneficial to implement, I believe that it will be worse for the average player and drive more players away from mod in favour of other supernatural mods e.g. bewitchment. I do however think that the activation could be implemented as a simple addition to the mod and would improve the modpack and users experiences as a small quality of life update.

**Comment by MarioSMB** [comment]
I understand that flight is a tender topic here, as it is afterall a very promiment aspect of the mod.  

With that said, balance is about finding a happy medium between power and utility, which is where Vampirism should sit arguably; it's providing a unique twist on gameplay with some special abilities.  
Making the player too stacked as a means of incentive to play this over other mods is the same approach mods with netherite-tier ores in the overworld follow, and it's sadly quite misguided. People will come here for the fun of playing as a vampire and all that encompasses, not because of the creative-mode flight.

There is more fun to be had in playing with a unique twist and the gameplay challenges that come with it, being overpowered takes away from that experience.  
Flight is locked behind the end-game for a reason; there are next to no threats in the sky, it's all on the ground. If you can just fly over the land during the night and swim the rest of the way during the day (infinite air and all) it trivializes a large portion of the game's challenge.

**Comment by Piklach** [comment]
>"Making the player too stacked as a means of incentive to play this over other mods is the same approach mods with netherite-tier ores in the overworld follow, and it's sadly quite misguided. People will come here for the fun of playing as a vampire and all that encompasses, not because of the creative-mode flight." 

I apologise if you misunderstand but I do not ask that we make the mod unreasonable to lure players in, I ask that we don't throw away users needlessly from this mod by making life harder upon them for no reason. I'm not making vampires too overpowered, we test the balance between Hunters and vampires and also werewolves from the addon mod, and we do this at a server level to put it through rigorous testing. To say that it is unbalanced simply because you prefer the idea of using a jetpack over the mod laid infront of you is negative towards the other users who love this mod the way that it is. 

>"People will come here for the fun of playing as a vampire and all that encompasses, not because of the creative-mode flight."

Every single component of the mod is put together to create a fair balance between the races and to allow players to have as much fun when coming to join and play it. Removing any one part in favour of something most would see as a negative, turns the entire aspect as a whole to diminish. 

>"If you can just fly over the land during the night and swim the rest of the way during the day (infinite air and all) it trivializes a large portion of the game's challenge". 

If we are considering minecrafts base mechanics on the ground to be a challenge, then who are the target audience, young kids (6-12years old), teenagers (13 to 18) or adults (18+). I understand you seem to think it removes some challenge being able to fly over these mobs and terrain, however, it just removes a mundane part of travel across the map leading players to spend more time walking and being pestered by mobs. (These mobs die relatively easily with two swings of a sword) So, how is forcing players to travel among them making the mod more interesting? It just slows down the user and causes them to undertake more of a farce and effort to travel the map in order to gain experience and the equipment required to level up. 

>"There is more fun to be had in playing with a unique twist and the gameplay challenges that come with it, being overpowered takes away from that experience." 

Possibly, however, the alternative you wish to use, a jetpack like system exists already. Many other mods implement this and if you believe that is the proper way to play, I suggest you add one of these mods to your modpack. Many users have differing opinions on this, and like you said, this is arguably a very tender topic. This being stated, I believe to suggest a change for the entire mod when there is already an alternative out there for you, is an insult to all the other players who prefer flight mode the way it is currently implemented.

**Comment by MarioSMB** [comment]
I seem to be getting personally targeted for my feedback here, so I'm going to leave it to the developers to decide how to continue with this.

**Comment by Boozooo** [comment]
No, you are merely someone with a take so bad it is hard to refrain from actual personal attacks - which nobody besides me did. Nobody attacked you, only your poor opinion. You probably would think it is too overpowered to have the speed buff from the skill as it prevents the challenge of being slow.

Pik worked hard on the response to you, and yet you dismiss it all as a personal attack. That explains a lot about your stance. Please actually answer normally next time, or just refrain from writing, you child.

**Comment by supermike1999** [comment]
dang, what an interesting comment at the end.

**Comment by maxanier** [comment]
Without further diving into the topic itself, I would like to state that there are very different audiences of the mod.
Some would like a "true to common vampire lore" experience, some want a godlike vampire experience. Some are very experienced with Minecraft and can defeat the ender dragon within an hour, others cannot (me included).
Some want an exciting adventure, others a creative challenge.

Vampirism currently doesn't focus on a narrow target audience which is good and bad. I would love to have different flavors or addons, but that is just too much work. Generally, we of course tend to go in the direction of our most loyal (and vocal) players.

Based on that, I do believe that this overhaul is a good improvement - for some - and a bad decision for others. IMHO It's less of a discussion on if this is a good choice than a discussion who this would work for.

**Comment by MarioSMB** [comment]
Looking back on this, I believe the real focus should be on the fact that all of this would be *possible*, even keeping the current behaviour, with the introduction of a custom flight system for bat mode.  
One would have the freedom to customise it as they please! Being able to fix (turn off) double-jump and sprinting as a bat would just be nice bonuses to this extension.

**Comment by kaboom24** [comment]
Personally I am not a big fan of this idea.  I feel like it removes one of my favorite aspects of vampirism. Vampires are famous for turning into bats and this would berg it so much that no one would use it anymore. After all,  vampire speed just about equals Bat mode and when you said it removes the challenge from the game most of the mobs just ignore you anyway so there is no difference between ground travel and bat mode except that ground travel is more boring. It would just be too much effort to use bay mode and not worth it. So I don't think it should be added

**Comment by MarioSMB** [comment]
A custom flight system would let the developers fix bugs and make improvements that are not possible now, doesn't mean they have to change bat mode or nerf it.  
There are no downsides to this, any "extra" features I suggested are just that - suggestions that could be considered when developing the system. 

I think the real discussion to be held is an internal one; that is to say, the developers are welcome to use this as a tracking issue but are free to decide when and how they want to implement a flight system, if at all. I'm just trying to help get the ball rolliing as I've encountered the issues that cannot be resolved thanks to creative flight limitations.  

As much as I'd love to see this done, I can understand that it's a huge task for seemingly little benefit (and a lot of community backlash so far), so as a long time player of this mod I am hopeful for the future and grateful for what we have so far!
